backend/modules/simulation/service.py

```
@@ -148,10 +162,47 @@ async def process_chat_message(
+                            uploaded_count = await sandbox_service.upload_scene_data_files(
+                                user_progress.sandbox_id, scene_data_files
+                            )
+                            if uploaded_count < expected_count:
+                                logger.error(
+                                    f"[SERVICE] Partial upload failure for scene {next_scene_id}: "
+                                    f"uploaded {uploaded_count}/{expected_count} files"
+                                )
+                                # Clear progression result to prevent advancing into broken code challenge
+                                progression_result['next_scene'] = None
+                                progression_result['next_scene_id'] = None
+                                raise HTTPException(
```

Ruff (0.15.7)

[error] Undefined name `HTTPException`

(F821)